advanced_source/cpp_extension.rst

```diff
@@ -0,0 +1,1033 @@
+
+Performance Comparison
+**********************
+
+Now that we are able to use and call our C++ code from PyTorch, we can run a
+small benchmark to see how much performance we gained from rewriting our op in
+C++. We'll run the LLTM forwards and backwards a few times and measure the
+duration::
+
+  import torch
+
+  batch_size = 16
+  input_features = 32
+  state_size = 128
+
+  X = torch.randn(batch_size, input_features)
+  h = torch.randn(batch_size, state_size)
+  C = torch.randn(batch_size, state_size)
+
+  rnn = LLTM(input_features, state_size)
+
+  forward = 0
+  backward = 0
+  for _ in range(100000):
+      start = time.time()
+      new_h, new_C = rnn(X, (h, C))
+      forward += time.time() - start
+
+      start = time.time()
+      (new_h.sum() + new_C.sum()).backward()
+      backward += time.time() - start
+
+  print('Forward: {:.3f} us | Backward {:.3f} us'.format(forward * 1e6/1e5, backward * 1e6/1e5))
+
+If we run this code with the original LLTM we wrote in pure Python at the start
+of this post, we get the following numbers (on my machine)::
+
+  Forward: 506.480 us | Backward 444.694 us
+
+and with our new C++ version::
+
+  Forward: 349.335 us | Backward 443.523 us
+
+We can already see a significant speedup for the forward function (more than
+30%). For the backward function a speedup is visible, albeit not major one. The
+backward pass I wrote above was not particularly optimized and could definitely
+be improved. Also, PyTorch's automatic differentiation engine can automatically
+parallelize computation graphs, may use a more efficient flow of operations
+overall, and is also implemented in C++, so it's expected to be fast.
+Nevertheless, this is a good start.
+
+Performance on GPU Devices
+**************************
+
+A wonderful fact about PyTorch's *ATen* backend is that it abstracts the
+computing device you are running on. This means the same code we wrote for CPU
+can *also* run on GPU, and individual operations will correspondingly dispatch
+to GPU-optimized implementations. For certain operations like matrix multiply
+(like ``mm`` or ``admm``), this is a big win. Let's take a look at how much
+performance we gain from running our C++ code with CUDA tensors. No changes to
+our implementation are required, we simply need to move our tensors to GPU
+memory with ``.cuda()`` from Python::
+
+  import torch
+
+  assert torch.cuda.is_available()
+
+  batch_size = 16
+  input_features = 32
+  state_size = 128
+
+  # Note the .cuda() calls here
+  X = torch.randn(batch_size, input_features).cuda()
+  h = torch.randn(batch_size, state_size).cuda()
+  C = torch.randn(batch_size, state_size).cuda()
+
+  rnn = LLTM(input_features, state_size).cuda()
+
+  forward = 0
+  backward = 0
+  for _ in range(100000):
+      start = time.time()
+      new_h, new_C = rnn(X, (h, C))
+      forward += time.time() - start
+
+      start = time.time()
+      (new_h.sum() + new_C.sum()).backward()
```

Comment:
Don't we need synchronization to get correct timings?

Reply:
what?